op-node/rollup/derive/blob_data_source.go

```diff
@@ -0,0 +1,131 @@
+package derive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+// BlobDataSource fetches both call-data (backup) and blobs and transforms them into usable rollup data.
+type BlobDataSource struct {
+	open           bool
+	callData       []eth.Data
+	blobDataHashes []eth.IndexedDataHash
+
+	blobs []*eth.Blob
+
+	ref         eth.L1BlockRef
+	batcherAddr common.Address
+
+	cfg *rollup.Config
+
+	fetcher      L1TransactionFetcher
+	blobsFetcher L1BlobsFetcher
+	log          log.Logger
+}
+
+// NewBlobDataSource creates a new blob-data source.
+func NewBlobDataSource(ctx context.Context, log log.Logger, cfg *rollup.Config, fetcher L1TransactionFetcher, blobsFetcher L1BlobsFetcher, ref eth.L1BlockRef, batcherAddr common.Address) DataIter {
+	return &BlobDataSource{
+		open:         false,
+		ref:          ref,
+		cfg:          cfg,
+		fetcher:      fetcher,
+		log:          log.New("origin", ref),
+		batcherAddr:  batcherAddr,
+		blobsFetcher: blobsFetcher,
+	}
+}
+
+// Next returns the next piece of data if any remains. It returns ResetError if it cannot find the
+// referenced block or a referenced blob, or TemporaryError for any other failure to fetch a block
+// or blob.
+func (ds *BlobDataSource) Next(ctx context.Context) (eth.Data, error) {
+	if !ds.open {
+		if _, txs, err := ds.fetcher.InfoAndTxsByHash(ctx, ds.ref.Hash); err == nil {
+			ds.open = true
+			ds.callData, ds.blobDataHashes = BlobDataFromEVMTransactions(ds.cfg, ds.batcherAddr, txs, ds.log)
+		} else if errors.Is(err, ethereum.NotFound) {
+			return nil, NewResetError(fmt.Errorf("failed to open blob-data source: %w", err))
+		} else {
+			return nil, NewTemporaryError(fmt.Errorf("failed to open blob-data source: %w", err))
+		}
+	}
+	// prioritize call-data
```

Comment:
This behavior needs to be very explicit in the specs and perhaps we want a sanity check in the batcher to make it very explicit across the codebase there that SHOULD only be calldata or blobs and blobs are ignored if calldata is present

It silently ignores the blobs here when calldata is present, if it doesn't add too much complexity to the implementation it would be nice to defensively log something.

I am not sure if we want to exist in a world where we can derive both calldata and blobs from the same transaction so i understand the approach here

Reply:
Blob data would not be ignored: if there was blobdata from the same transaction it would be returned by a later call to Next().  See the implementation in BlobDataFromEVMTransactions below which is where both calldata and blob data hashes are extracted (possibly from a single tx). That said the batcher will not (initially) be using both calldata and blob data, but we'd have the flexibility to add that later if it turned out to be worth doing at some point.

Reply:
I would like the exact behavior here documented in the specs so that we can match the specified behavior against the implementation before merging this to ensure there are no ambiguities

Reply:
Based on call today we'll change this so that if you post as a Blob then calldata can NOT be used for batches (so it's either or on a per TX level).  @protolambda can you put this in your Spec PR?

Reply:
Updated the derivation code so that calldata in blob transactions is ignored.

Reply:
Spec change: https://github.com/ethereum-optimism/optimism/pull/8660